core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java

```
@@ -29,16 +44,56 @@
+	 */
+	public OAuth2TokenEndpointFilter(OAuth2AuthorizationService authorizationService) {
+		Assert.notNull(authorizationService, "authorizationService cannot be null");
+		this.authorizationGrantConverter = new OAuth2AuthorizationCodeAuthenticationTokenConverter();
+		this.authenticationManager = new OAuth2ClientAuthenticationManager();
+		this.uriMatcher = new AntPathRequestMatcher(DEFAULT_TOKEN_URI, HttpMethod.POST.name());
+		this.authorizationService = authorizationService;
+	}
 
 	@Override
 	protected void doFilterInternal(HttpServletRequest request,
 			HttpServletResponse response, FilterChain filterChain)
 			throws ServletException, IOException {
+		if (uriMatcher.matches(request)) {
+			String errorCode = validateAccessTokenRequest(request);
+			if (errorCode == null) {
+				OAuth2AuthorizationCodeAuthenticationToken authCodeAuthToken = authorizationGrantConverter.convert(request);
+				OAuth2AccessTokenAuthenticationToken accessTokenAuthenticationToken =
+						(OAuth2AccessTokenAuthenticationToken) authenticationManager.authenticate(authCodeAuthToken);
+				OAuth2Authorization authorization = authorizationService
+						.findByTokenAndTokenType(authCodeAuthToken.getCode(), TokenType.AUTHORIZATION_CODE);
+				authorization.setAccessToken(accessTokenAuthenticationToken.getAccessToken());
+				authorizationService.save(authorization);
+			} else {
+				throw new OAuth2AuthenticationException(new OAuth2Error(errorCode));
+			}
+		}
+		filterChain.doFilter(request, response);
+	}
 
+	private String validateAccessTokenRequest(HttpServletRequest request) {
+		if (!AuthorizationGrantType.AUTHORIZATION_CODE.getValue().equals(request.getParameter(OAuth2ParameterNames.GRANT_TYPE))) {
+			return OAuth2ErrorCodes.UNSUPPORTED_GRANT_TYPE;
+		} else if (StringUtils.isEmpty(request.getParameter(OAuth2ParameterNames.CODE))
+				|| StringUtils.isEmpty(request.getParameter(OAuth2ParameterNames.REDIRECT_URI))) {
+			return OAuth2ErrorCodes.INVALID_REQUEST;
```

Instead of returning `String`, just `throw OAuth2AuthenticationException`

---

@jgrandja and should the exception be just thrown or also caught and written to the response with `response.sendError`?

---

The `OAuth2Error` should also be written to the response.